src/com/amazon/ion/impl/IonReaderBinaryIncremental.java

```diff
@@ -0,0 +1,1932 @@
+        new _Private_RecyclingStack.ElementFactory<ContainerInfo>() {
+
+        @Override
+        public ContainerInfo newElement() {
+            return new ContainerInfo();
+        }
+    };
+
+    // The Ion 1.0 system symbol table.
+    private static final List<String> SYSTEM_SYMBOLS_1_0 = Collections.unmodifiableList(new ArrayList<String>() {{
+        add(null);
+        add("$ion");
+        add("$ion_1_0");
+        add("$ion_symbol_table");
+        add("name");
+        add("version");
+        add("imports");
+        add("symbols");
+        add("max_id");
+        add("$ion_shared_symbol_table");
+    }});
+
+    // The size of the Ion 1.0 system symbol table.
+    private static final int SYSTEM_SYMBOLS_1_0_SIZE = SYSTEM_SYMBOLS_1_0.size();
+
+    // The system symbol table SID for the text "$ion_symbol_table".
+    private static final int ION_SYMBOL_TABLE_ID = 3;
+
+    // The system symbol table SID for the text "name".
+    private static final int NAME_ID = 4;
+
+    // The system symbol table SID for the text "version".
+    private static final int VERSION_ID = 5;
+
+    // The system symbol table SID for the text "imports".
+    private static final int IMPORTS_ID = 6;
+
+    // The system symbol table SID for the text "symbols".
+    private static final int SYMBOLS_ID = 7;
+
+    // The system symbol table SID for the text "max_id".
+    private static final int MAX_ID_ID = 8;
+
+    // The size of the reusable UTF-8 decoding buffer.
+    private static final int UTF8_BUFFER_SIZE_IN_BYTES = 4 * 1024;
+
+    // The final byte of the binary IVM.
+    private static final int IVM_FINAL_BYTE = 0xEA;
+
+    // Isolates the highest bit in a byte.
+    private static final int HIGHEST_BIT_BITMASK = 0x80;
+
+    // Isolates the lowest seven bits in a byte.
+    private static final int LOWER_SEVEN_BITS_BITMASK = 0x7F;
+
+    // Isolates the lowest six bits in a byte.
+    private static final int LOWER_SIX_BITS_BITMASK = 0x3F;
+
+    // The number of significant bits in each UInt byte.
+    private static final int VALUE_BITS_PER_UINT_BYTE = 8;
+
+    // The number of significant bits in each VarUInt byte.
+    private static final int VALUE_BITS_PER_VARUINT_BYTE = 7;
+
+    // An IonCatalog containing zero shared symbol tables.
+    private static final IonCatalog EMPTY_CATALOG = new SimpleCatalog();
+
+    // Initial capacity of the stack used to hold ContainerInfo. Each additional level of nesting in the data requires
+    // a new ContainerInfo. Depths greater than 8 will be rare.
+    private static final int CONTAINER_STACK_INITIAL_CAPACITY = 8;
+
+    // Initial capacity of the ArrayList used to hold the symbol IDs of the annotations on the current value.
+    private static final int ANNOTATIONS_LIST_INITIAL_CAPACITY = 8;
+
+    // Initial capacity of the ArrayList used to hold the text in the current symbol table.
+    private static final int SYMBOLS_LIST_INITIAL_CAPACITY = 128;
```

Comment:
I'm not sure I understand the concern, can you clarify? When I create ArrayLists and know that the expected size is larger than the default size, I specify a higher size. That's what I've done here. In this case, 128 seems like a good number that will allow many small streams never to have to grow, while larger streams will have to grow fewer times. The performance difference is probably small, but the effort to get there is trivial.

> The reference copying and allocation (in that case) is a trivial compared to any of the buffer copying and moving we're doing here

What buffer copying and moving are we doing here?

Reply:
> What buffer copying and moving are we doing here?

In this stack of Reader => Buffer => InputStream we're doing a lot of byte copying and some number of allocations. By specifying a certain initial size for this ArrayList we're saving a certain amount of allocation and byte copying (reference copies, when the backing array is grown), but it's a fixed amount and not an improvement to scaling overhead.

I was trying to say that it just doesn't seem worth it to have a commented constant for this. We don't use the number anywhere else, I don't know any reason to expect that 128 is the "right" size, and the cost of the number of grow operations required to go from the default size to 128 or greater is trivial. If we need to grow the symbol table from the default ArrayList size to anything >= 128 then starting at 128 saved us... six allocations and ~390 reference copies? (10 + 10*(1.6) + 10*(1.6**2) ... - (160 - 128)).

If we start with 128 and reach 129 symbols we have two allocations and 128 copied references. If we reach 205 symbols we have three allocations and 333 copied references. 

205 symbols starting from the default ArrayList capacity (of 10) is something like seven allocations and 580 copied references? Six allocations and 420 reference copies to get us to capacity 160, and an additional allocation and 160 reference copies to get us to a container than holds > 205. 

Maybe my math is wrong but the general shape stays the same. I don't see a reason to worry about these details. When positioned next to constants like LONG_SIZE_IN_BYTES I see SYMBOLS_LIST_INITIAL_CAPACITY and it makes me start thinking "why 128? what am I missing here?". 

YMMV, but that's my thought process.

Reply:
> I don't know any reason to expect that 128 is the "right" size

We don't know 128 is the "right" size (there is no one-size-fits-all for all possible Ion streams), but we *do* know that the default of 10 is the *wrong* size. That's because the first thing we do to the `symbols` list after instantiating it is add 10 system symbols to it. So we've guaranteed a growth if the stream contains even one user symbol, which almost all will. Because of that, I don't see a good reason not to raise the initial size. If you agree with that, then now we're just negotiating the amount. I agree that it's a small benefit, but it's also a zero-effort benefit (not including the time it's taken to discuss it).

I think one thing I may still not understand is whether you have a problem with specifying 128 specifically, or specifying a value at all, or simply with creating a constant for the selected value. Can you clarify?